src/main/scala/io/computenode/cyfra/dsl/GFunction.scala

```diff
@@ -8,18 +8,24 @@ import izumi.reflect.Tag
 
 import scala.deriving.Mirror
 
-case class GFunction[H <: Value : Tag: FromExpr, R <: Value : Tag : FromExpr](fn: H => R)(implicit context: GContext){
-  def arrayInputs: List[Tag[_]] = List(summon[Tag[H]])
-  def arrayOutputs: List[Tag[_]] = List(summon[Tag[R]])
-  val pipeline: ComputePipeline = context.compile(this)
-}
+// case class GFunction[H <: Value : Tag: FromExpr, R <: Value : Tag : FromExpr](fn: H => R)(implicit context: GContext){
+//   def arrayInputs: List[Tag[_]] = List(summon[Tag[H]])
+//   def arrayOutputs: List[Tag[_]] = List(summon[Tag[R]])
+//   val pipeline: ComputePipeline = context.compile(this)
+// }
 
-case class GArray2DFunction[G <: GStruct[G] : GStructSchema : Tag, H <: Value : Tag : FromExpr, R <: Value : Tag : FromExpr](
+case class GFunction[G <: GStruct[G] : GStructSchema : Tag, H <: Value : Tag : FromExpr, R <: Value : Tag : FromExpr](
   width: Int,
   height: Int,
   fn: (G, (Int32, Int32), GArray2D[H]) => R
-)(implicit context: GContext){
+)(implicit context: GContext){ // add uniformContext here?
```

Comment:
It should be in the apply, not constructor. Function can be compiled once and then run multiple time in various uniform contexts. For example  one function is executed multiple times when animation is being rendered, but `time` value changes in an uniform.